File: FindWeather/FindWeather v112/FindWeather v112/Adapters/OpenWeatherAdapter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace FindWeather_v112
{
    internal class OpenWeatherAdapter:IWeatherProvider
    {
        public OpenWeatherAPI baseClient = new OpenWeatherAPI();
        public async Task<WeatherData> GetWeatherAsync(double latitude, double longitude)
        {
            var response = await baseClient.FetchWeatherDataByCoordinates(latitude, longitude);
            var json = JsonDocument.Parse(response);
            return new WeatherData
            {
                City = json.RootElement.GetProperty("name").GetString(),
                Temperature = json.RootElement.GetProperty("main").GetProperty("temp").GetDouble(),
                Condition = json.RootElement.GetProperty("weather")[0].GetProperty("description").GetString(),
                Source = "OpenWeatherMap"
            };
        }
        public async Task<WeatherData> GetWeatherByCityAsync(string city)
        {
            var response = await baseClient.FetchWeatherDataByCity(city);
            var json = JsonDocument.Parse(response);
            return new WeatherData//same ase the previous one!
            {
                City = json.RootElement.GetProperty("name").GetString(),
                Temperature = json.RootElement.GetProperty("main").GetProperty("temp").GetDouble(),
                Condition = json.RootElement.GetProperty("weather")[0].GetProperty("description").GetString(),
                Source = "OpenWeatherMap"
            };
        }

    }
}

File: FindWeather/adapter/OpenWeatherAdapter.java
package adapter;

import api.OpenWeatherAPI;
import models.WeatherData;
import org.json.JSONObject;

public class OpenWeatherAdapter implements WeatherProvider {
    private final OpenWeatherAPI api;

    public OpenWeatherAdapter() {
        this.api = new OpenWeatherAPI();
    }

    @Override
    public WeatherData getWeatherByIP(double latitude, double longitude) throws Exception {
        JSONObject response = api.getWeatherByIP(latitude, longitude);
        return parseWeatherData(response, "Location via IP");
    }

    @Override
    public WeatherData getWeatherByCity(String city) throws Exception {
        JSONObject response = api.getWeatherByCity(city);
        return parseWeatherData(response, city);
    }

    private WeatherData parseWeatherData(JSONObject response, String locationName) throws Exception {
        if (!response.has("main") || !response.has("weather")) {
            throw new Exception("Invalid response from OpenWeather API");
        }

        double temperature = response.getJSONObject("main").getDouble("temp");
        String conditions = response.getJSONArray("weather").getJSONObject(0).getString("description");

        return new WeatherData(locationName, temperature, conditions, "OpenWeather");
    }
}
